src/VariablesState.ts

```diff
@@ -0,0 +1,311 @@
+
+export class VariablesState{
+	// The way variableChangedEvent is a bit different than the reference implementation.
+	// Originally it uses the C# += operator to add delegates, but in js we need to maintain
+	// an actual collection of delegates (ie. callbacks) to register a new one, there is a
+	// special ObserveVariableChange method below.
+	public variableChangedEventCallbacks: Array<(variableName: string, newValue: InkObject) => void> = [];
+	public variableChangedEvent(variableName: string, newValue: InkObject): void {
+		this.variableChangedEventCallbacks.forEach((callback) => {
+			callback(variableName, newValue);
+		});
+	}
+
+	get batchObservingVariableChanges(){
+		return this._batchObservingVariableChanges;
+	}
+	set batchObservingVariableChanges(value: boolean){
+		value = !!value; // TODO: Might have become useless with TypeScript.
+		this._batchObservingVariableChanges = value;
+		if (value) {
+			this._changedVariables = new Set();
+		}
+
+		else {
+			if (this._changedVariables != null) {
+				for (let variableName of this._changedVariables) {
+					let currentValue = this._globalVariables.get(variableName);
+					if (!currentValue) {
+						throwNullException('currentValue');
+					} else {
+						this.variableChangedEvent(variableName, currentValue);
+					}
+				}
+			}
+		}
+	}
+
+	get callStack(){
+		return this._callStack;
+	}
+	set callStack(callStack: any /* CallStack */){
+		this._callStack = callStack;
+	}
+
+	private _batchObservingVariableChanges: boolean = false;
+
+	// the original code uses a magic getter and setter for global variables,
+	// allowing things like variableState['varname]. This is not quite possible
+	// in js without a Proxy, so it is replaced with this $ function.
+	public $(variableName: string, value: InkObject){
+		if (typeof value === 'undefined'){
+			let varContents = this._globalVariables.get(variableName);
+
+			if ( typeof varContents === 'undefined' ) {
+				varContents = this._defaultGlobalVariables.get(variableName);
+			}
+
+			if ( typeof varContents !== 'undefined' )
+				return (varContents as AbstractValue).valueObject;
+			else
+				return null;
+		}
+		else{
+			if (typeof this._defaultGlobalVariables.get(variableName) === 'undefined')
+				throw new StoryException("Cannot assign to a variable that hasn't been declared in the story");
+
+			let val = Value.Create(value);
+			if (val == null) {
+				if (value == null) {
+					throw new StoryException('Cannot pass null to VariableState');
+				} else {
+					throw new StoryException('Invalid value passed to VariableState: '+value.toString());
+				}
+			}
+
+			this.SetGlobal(variableName, val);
+		}
+	}
+
+	constructor(callStack: any /* CallStack */, listDefsOrigin: ListDefinitionsOrigin){
+		this._globalVariables = new Map();
+		this._callStack = callStack;
+		this._listDefsOrigin = listDefsOrigin;
+
+		// TODO should we discourage the use of proxies since they're non-portable
```

Comment:
Can you elaborate more on this please?

Reply:
I don't know much about proxies, this is most likely a comment inserted by @NQNStudios, so you'd have to ask them!